src/main/Main.hs

```diff
@@ -470,21 +473,41 @@ setupCmd SetupCmdOpts{..} go@GlobalOpts{..} = do
               )
           Nothing
 
-withConfig :: GlobalOpts
+-- | Enforce mutual exclusion of every action running via this function
+-- on this users account.  Currently, stack uses this to completely
+-- exclude concurrent execution of stack commands.  In the future,
+-- stack may refine this to a finer-grained locking approach.
+withUserFileLock :: Config
+                 -> (FileLock -> IO a)
+                 -> IO a
+withUserFileLock cfg act = do
+    lockfile <- parseRelFile "lockfile"
+    let pth = configStackRoot cfg </> lockfile
+    fstTry <- tryLockFile (toFilePath pth) Exclusive
```

Comment:
Ah, good catch, thanks.

Ok, let's see.  So if anything goes really wrong and the process exits, then the `filelock` lib is designed so that file descriptors get closed and the locks get released.

But the issue here is that if the thread running `withUserFileLock` gets killed by an async exception from another Haskell IO thread -- while holding the lock from `tryLockFile` -- then the lock will not get released until the process exits.

I think this is ok for now because the lock is global, and per-process.  But it's better to fix this and use bracket at the outset.  Ok, just did a commit with that fix.